Support shell expansion of filenames from wgetrc

darnir opened the pull request:
* bootstrap.conf: Add module wordexp
* src/options.c (_shell_expand): New function to perform local shell
expansion on a given string
(parse_filename): Parse the input string and perform shell expansion on
it using _shell_expand
(parse_string): Set the parser functions for various options that may be
set from a config file to use parse_filename

darnir commented:
This should mostly finish #132. Only thing pending would be documenting these features

rockdaboot commented:
See my comment  on #132 

With latest gnulib:
```
gnulib-tool: warning: module wordexp doesn't exist
```

rockdaboot commented:
It's a POSIX function, so at least on Windows we have a problem.

rockdaboot commented:
```
if (*src == '~')
  expanded_str = wget_str_asprintf("%s%s", get_home_dir(), src + 1);
else
  expanded_str = wget_strdup(src);
```

darnir commented:
@rockdaboot  I've changed the commit to use `glob()` instead of `wordexp()`. 
This is a WIP commit. I wrote it quickly before class in the morning and there might be possibilities for optimization. This will also cause valgrind to report reachable blocks (No memory leaks).

If you're okay with this, I can also split the function `wget_strnglob` into libwget/utils.c and export it from there. Else, I'll rename it so as not to use the libwget namespace. (Of course, if we move it to utils, even `get_home_dir()`should use it then)

rockdaboot commented:
@Darshit After reading up a few things and looking at bash's tilde expansion C code, I suggest that we leave tilde expansion to the shell. Each shell has it's special cases when to expand tilde. There is also tilde expansion to $PWD and $OLDPWD.

That means e.g. `--config ~/.wgetrc` works while `--config=~/.wgetrc` doesn't (tilde in latter example won't be expanded). We should document this.

If in doubt, people should use $HOME (or $PWD or whatever) instead of using tilde.

Adding tilde expansion to wget2 would introduce incompatibilities and quirks that we can well do without.

darnir commented:
@rockdaboot While I in general agree that the tilde expansion should be left to the shell, one very important use case is supporting tilde expansion for files passed from `wgetrc`. Having a hard coded absolute path in the wgetrc should not be a requirement. I don't say we support all forms of tilde expansion, but *atleast* when the command comes from a wgetrc file, we should perform expansion into the $HOME directory. It can be documented that only this feature is available.

This is why I would have liked to use `wordexp` it uses the underlying shell to do the expansions. 
When arguments are passed using the command line, the shell performs whatever expansions are required, so that will always work on all platforms where Wget2 can execute.

However, when reading from a config file, we can document that tilde and shell expansions are only done on supported platforms. Apart from Windows, `wordexp` is available in most recent operating systems. GNULib lists these as platforms with portability issues:
 Mac OS X 10.3, OpenBSD 3.8, Minix 3.1.8, IRIX 5.3, Cygwin 1.5.x, mingw, MSVC 9, Interix 3.5, BeOS. 

They are mostly all 10+ years old. Wasn't one of our goals with Wget2 to target newer platforms and let Wget service the old ones? 

There may be quirks or incompatibilities, but I strongly believe that allowing tilde-filenames inside `wgetrc` is a requirement that we cannot relax.

rockdaboot commented:
OK. Then we have to document, that we only tilde-expand files from config files and only in a certain way (e.g. point to 'man wordexp' or 'man glob' - whatever you decide for).

So I leave it to you if you prefer wordexp or glob.

But wget_strnglob() curretnly is not reentrant and thus no option for libwget.

darnir commented:
As much as I'd like to use `wordexp` for its simplicity, I think I'll settle for `glob` mostly because it is far more portable.

A reentrant version of wget_strnglob would use no static variables and we can drop the GLOB_APPEND flag.

darnir commented on src/options.c:
```diff
@@ -247,38 +247,18 @@ static int G_GNUC_WGET_NORETURN print_help(G_GNUC_WGET_UNUSED option_t opt, G_GN
-		expanded_str = wget_str_asprintf("%s%s", homestr, pathptr+1);
-		xfree(homestr);
-	} else {
+		expanded_str = wget_strnglob(str, pathptr - str, GLOB_TILDE|GLOB_ONLYDIR|GLOB_NOCHECK);
```
You are right, I missed that. I've pushed a fix. Once everything looks good, I will squash, write a clean commit message and push

rockdaboot replied:
Maybe you can squash it right now, makes review easier on Github.
I'll have a closer look at it tomorrow...

rockdaboot commented:
I am fine with your patch, just add 'strchrnul' to bootstrap.conf.

rockdaboot commented:
Any progress here ? strchrnul meanwhile is in bootstrap.conf, src/options.c needs small changes.

darnir commented:
I am on vacation for a week. Needs a very small set of changes which I'll
finish by the end of the week.

Thanking You,
Darshit Shah
Sent from mobile device. Please excuse my brevity

On 17-Dec-2016 9:32 PM, "Tim Rühsen" <notifications@github.com> wrote:

> Any progress here ? strchrnul meanwhile is in bootstrap.conf,
> src/options.c needs small changes.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/rockdaboot/wget2/pull/134#issuecomment-267785059>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/ABlUtSPLCOpd-XJ8pbHsfezu55kfnhpxks5rJEbYgaJpZM4LBmXg>
> .
>

rockdaboot commented:
Thanks Darshit. Slightly amended and pushed.
